kbfscrypto/encrypted_data.go

```diff
@@ -67,37 +86,42 @@ func encryptData(data []byte, key [32]byte) (encryptedData, error) {
 		return encryptedData{}, err
 	}
 
-	sealedData := secretbox.Seal(nil, data, &nonce, &key)
+	return encryptDataWithNonce(data, key, nonce, EncryptionSecretbox)
+}
 
-	return encryptedData{
-		Version:       EncryptionSecretbox,
-		Nonce:         nonce[:],
-		EncryptedData: sealedData,
-	}, nil
+// decryptDataWithNonce decrypts the given encrypted data with the
+// given symmetric key and nonce.
+func decryptDataWithNonce(
+	encryptedData encryptedData, key [32]byte, nonce [24]byte) ([]byte, error) {
+	decryptedData, ok := secretbox.Open(
+		nil, encryptedData.EncryptedData, &nonce, &key)
+	if !ok {
+		return nil, errors.WithStack(libkb.DecryptionError{})
+	}
+
+	return decryptedData, nil
 }
 
 // decryptData decrypts the given encrypted data with the given
 // symmetric key.
 func decryptData(encryptedData encryptedData, key [32]byte) ([]byte, error) {
-	if encryptedData.Version != EncryptionSecretbox {
+	switch encryptedData.Version {
+	case EncryptionSecretbox, EncryptionSecretboxWithKeyNonce:
+	// We're good.
+	default:
 		return nil, errors.WithStack(
 			UnknownEncryptionVer{encryptedData.Version})
 	}
 
+	// For v2, the caller must have already checked the nonce.
```

Comment:
Is there any chance we'd lose this guarantee after a refactor?

Reply:
It's possible, though at least this is an unexported function, so the refactor would have to be within this file/package.

I split the code this way to avoid having a useless comparison check on the v1 code path, while still reusing code where possible between the two paths.  But happy to revisit that if you think it's too risky!

Reply:
can't you put the nonce check in the switch/case above? you'd have to pass in a dummy value for V1, but that doesn't seem so bad

Reply:
I could, but that's exactly what I was saying I wanted to avoid.  That check is useless for v1 and would just slow things down.  I'm not sure I see the benefit in forcing the check into this unexported function.

Would it be better if I renamed the function `decryptDataNoNonceCheck` or something more explicit?

Reply:
Ok @akalin-keybase cleared up my confusion, and my latest commit 4045666 moves the nonce check into `decryptData`.  It means we have to duplicate the call to get a 24-byte nonce in 4 different places, but I guess it's a bit cleaner.